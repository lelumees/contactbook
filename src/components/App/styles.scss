@import 'styles/main.scss';

.App {
    text-align: center;
    color: $primary-color;
}

.App-logo {
    height: 40px;
}

.App-header {
    height: 40px;
    padding: 20px 0 10px;
}

.App-title {
    color: $primary-color;
}

.App-searchbar {
    position: relative;
    box-sizing: border-box;
    width: 300px;
    margin: 0 auto;
}

.App-searchbar-input {
    width: 100%;
    box-sizing: border-box;
    background-color: lighten( $primary-color, 30% );
    color: darken( $primary-color, 10% );
    border: none;
    border-radius: 30px;
    padding: 12px 20px;
    font-size: 16px;
    font-weight: 700;
    transition: all 0.3s;
    text-transform: uppercase;

    &:focus {
        background-color: lighten( $primary-color, 25% );
    }
}

.App-helptext {
    font-size: 12px;
    font-weight: 700;
    color: $primary-color;
}

.App-contacts {
    max-width: 500px;
    margin: 20px auto;
}

.App-contact {
    padding: 16px 32px;
    background-color: lighten( $primary-color, 30% );
    border-radius: 8px;
    margin: 10px 0;
}

.App-contact__content {
    display: flex;
    justify-content: space-between;
}

.App-contact__actions {
    width: 100px;
}

.App-contact__data {
    line-height: 32px;
    text-transform: uppercase;
    font-size: 14px;
}

.App-button {
    display: inline-block;
    padding: 8px 16px;
    background: $primary-color;
    border-radius: 8px;
    color: #fff;
    font-weight: 700;
    text-transform: uppercase;
    font-size: 12px;
    cursor: pointer;

    &:hover {
        opacity: 0.8;
    }

    &.outlined {
        color: $primary-color;
        background: lighten( $primary-color, 33% );
        border: 1px solid $primary-color;
    }
}

.App-message-area {
    resize: none;
    border: none;
    margin: 16px 0 8px;
    width: 100%;
    padding: 12px;
    box-sizing: border-box;
    background: lighten( $primary-color, 25% );
    border-radius: 8px;
    height: 80px;
    color: $primary-color;
    font-weight: 700;
    font-size: 12px;
    transition: all 0.3s;

    &:focus {
        background: lighten( $primary-color, 23% );
    }
}

.App-message-actions {
    display: flex;
    justify-content: center;

    .App-button + .App-button {
        margin-left: 8px;
    }
}

.App-alert {
    position: fixed;
    bottom: 30px;
    left: 50%;
    width: 300px;
    margin-left: -150px;
    background: $bg-color;
    color: $primary-color;
    padding: 16px 32px;
    border-radius: 8px;

    &.error {
        background: $error-bg-color;
        color: $error-color;
    }

    &.success {
        background: $success-bg-color;
        color: $success-color;
    }
}
