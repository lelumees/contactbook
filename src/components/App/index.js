import React, { Component } from 'react';
import axios from 'axios';
import config from '../../config';
import logo from '../../assets/logo.svg';
import './styles.css';

const SEARCH_TIMEOUT_MS = 300;
let searchTimeout;

class App extends Component {

    constructor(props){
        super(props);

        this.state = {
            searchQuery: '',
            contacts: [],
            fetching: false,
            error: false,
            selectedContact: null,
            message: ''
        }

        this.handleSearchInputChange = this.handleSearchInputChange.bind(this);
        this.searchContacts = this.searchContacts.bind(this);
        this.handleSearchSuccess = this.handleSearchSuccess.bind(this);
        this.handleSearchError = this.handleSearchError.bind(this);
        this.renderHelpText = this.renderHelpText.bind(this);
        this.renderContact = this.renderContact.bind(this);
        this.handleMessageChange = this.handleMessageChange.bind(this);
        this.handleMessageClick = this.handleMessageClick.bind(this);
        this.handleCancelClick = this.handleCancelClick.bind(this);
    }

    handleSearchSuccess(response) {
        this.setState({ contacts: response.data, fetching: false });
    }

    handleSearchError(error) {
        console.error(error);
        this.setState({ error: true, fetching: false });
    }

    searchContacts() {
        const { searchQuery } = this.state;

        axios.get(config.contactsUrl, { params: { q: searchQuery } })
            .then(this.handleSearchSuccess)
            .catch(this.handleSearchError);
    }

    handleSearchInputChange(e) {
        clearTimeout(searchTimeout);

        const searchQuery = e.target.value;

        if (searchQuery.length < 2) {
            this.setState({ searchQuery, contacts: [], error: false, fetching: false, selectedContact: null });
            return;
        }

        this.setState({ searchQuery, fetching: true, error: false, selectedContact: null });
        searchTimeout = setTimeout(this.searchContacts, SEARCH_TIMEOUT_MS);
    }

    handleContactClick(contact) {
        this.setState({ selectedContact: contact });
    }

    handleMessageChange(e) {
        this.setState({ message: e.target.value });
    }

    handleMessageClick() {
        const { message, selectedContact } = this.state;

        if (!message || !message.length) return;
        this.setState({ message: '', selectedContact: null });

        axios.post(config.sendSMSUrl, { to: selectedContact.mobile, message })
            .then(result => console.log(result))
            .catch(err => console.error(err));
    }

    handleCancelClick() {
        this.setState({ message: '', selectedContact: null });
    }

    renderHelpText() {
        const { searchQuery, contacts, fetching, error } = this.state;
        if (error) return 'Oops, something went wrong with the request';
        if (fetching) return 'Searching...';
        if (searchQuery.length < 2) return 'Type at least 2 characters/numbers to start searching';
        if (!contacts.length) return 'Sorry, I could not find anyone';

        const countString = contacts.length === 1 ? 'contact' : 'contacts';
        return `Found ${contacts.length} ${countString}`;
    }

    renderContact(contact, index) {
        const { selectedContact, message } = this.state;
        const isSelected = selectedContact && selectedContact.id === contact.id;

        return (
            <div key={index} className='App-contact'>
                <div className='App-contact__content'>
                    <div className='App-contact__data strong'>{contact.name}</div>
                    <div className='App-contact__data'>{contact.mobile}</div>
                    <div className='App-contact__actions'>
                        { contact.mobile && !isSelected &&
                            <div className='App-button' onClick={ () => { this.handleContactClick(contact)} }>
                                Message
                            </div>
                        }
                    </div>
                </div>
                { isSelected &&
                    <div className=''>
                        <textarea className='App-message-area' onChange={this.handleMessageChange}>{message}</textarea>
                        <div className='App-message-actions'>
                            <div className='App-button outlined' onClick={this.handleCancelClick}>Cancel</div>
                            <div className='App-button' onClick={this.handleMessageClick}>Send SMS</div>
                        </div>
                    </div>
                }
            </div>
        );
    }

    render() {
        const { searchQuery, contacts } = this.state;

        return (
            <div className='App'>
                <header className='App-header'>
                    <img src={logo} className='App-logo' alt='logo' />
                </header>
                <h2 className='App-title'>Welcome to ContactBook!</h2>
                <p>Search for contacts and send them messages with ease.</p>
                <div className='App-searchbar'>
                    <input
                        type='text'
                        className='App-searchbar-input'
                        value={searchQuery}
                        onChange={this.handleSearchInputChange}
                    />
                </div>
                <p className='App-helptext'>{ this.renderHelpText() }</p>
                <div className='App-contacts'>{ contacts.map(this.renderContact) }</div>
            </div>
        );
    }
}

export default App;
