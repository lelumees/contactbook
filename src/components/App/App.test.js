import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import App from './index.js';

describe('App Snapshot', () => {
    it('renders', () => {
        const component = renderer.create(<App />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('App Component', () => {
    it('renders without crashing', () => {
        shallow(<App />);
    });

    it('renders welcome message', () => {
        const wrapper = shallow(<App />);
        const welcome = <h2 className='App-title'>Welcome to ContactBook!</h2>;
        expect(wrapper.contains(welcome)).toEqual(true);
    });

    it('simulates search event', () => {
        const handleChangeSpy = sinon.spy(App.prototype, 'handleSearchInputChange');
        const wrapper = mount(<App />);
        const event = {target: {name: 'App-searchbar-input', value: 'somequery'}};

        wrapper.find('.App-searchbar-input').simulate('change', event);
        expect(handleChangeSpy.calledOnce).toEqual(true);
    });
});
