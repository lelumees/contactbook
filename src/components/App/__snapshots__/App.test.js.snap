// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot renders 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
  </header>
  <h2
    className="App-title"
  >
    Welcome to ContactBook!
  </h2>
  <p>
    Search for contacts and send them messages with ease.
  </p>
  <div
    className="App-searchbar"
  >
    <input
      className="App-searchbar-input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <p
    className="App-helptext"
  >
    Type at least 2 characters/numbers to start searching
  </p>
  <div
    className="App-contacts"
  />
</div>
`;
